src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,237 @@
+            var trivia = root.GetLeadingTrivia();
+            var onBlankLine = false;
+            var inCopyright = true;
+            var removalList = new List<int>();
+            var leadingSpaces = string.Empty;
+            var possibleLeadingSpaces = string.Empty;
+
+            // Need to do this with index so we get the line endings correct.
+            for (var i = 0; i < trivia.Count; i++)
+            {
+                var triviaLine = trivia[i];
+                if (triviaLine.RawKind == FileHeaderHelper.SingleLineCommentTriviaKind)
+                {
+                    if (possibleLeadingSpaces != string.Empty)
+                    {
+                        leadingSpaces = possibleLeadingSpaces;
+                    }
+
+                    removalList.Add(i);
+                    onBlankLine = false;
+                }
+                else if (triviaLine.RawKind == FileHeaderHelper.WhitespaceTriviaKind)
+                {
+                    if (leadingSpaces == string.Empty)
+                    {
+                        possibleLeadingSpaces = triviaLine.ToFullString();
+                    }
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+                }
+                else if (triviaLine.RawKind == FileHeaderHelper.EndOfLineTriviaKind)
+                {
+                    possibleLeadingSpaces = string.Empty;
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+
+                    if (onBlankLine)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        onBlankLine = true;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            // Remove copyright lines in reverse order.
+            for (var i = removalList.Count - 1; i >= 0; i--)
+            {
+                trivia = trivia.RemoveAt(removalList[i]);
+            }
+
+            var newLineText = document.Project.Solution.Workspace.Options.GetOption(FormattingOptions.NewLine, root.Language);
+            var newLineTrivia = EndOfLine(newLineText);
+
+            var newHeaderTrivia = CreateNewHeader(leadingSpaces + FileHeaderHelper.CommentPrefix, expectedFileHeader, newLineText);
+
+            // Add a blank line after the header.
+            newHeaderTrivia = newHeaderTrivia.Add(newLineTrivia);
+
+            // Insert header at top of the file.
+            return root.WithLeadingTrivia(newHeaderTrivia.Add(newLineTrivia).AddRange(trivia));
+        }
+
+        private SyntaxNode AddHeader(Document document, SyntaxNode root, string expectedFileHeader)
+        {
+            var newLineText = document.Project.Solution.Workspace.Options.GetOption(FormattingOptions.NewLine, root.Language);
```

Comment:
aren't options exposed off of Solutoin as well?  that way we get a snapshot instead of a live-mutating version?

Reply:
fixed